util/aws_retryer.go

```diff
@@ -0,0 +1,135 @@
+package util
+
+import (
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+)
+
+type ConvoyAWSRetryer struct {
+	NumMaxRetries    int
+	MinDelay         int
+	MinThrottleDelay int
+	MaxDelay         int
+}
+
+// DefaultConvoyAWSRetryer creates a new ConvoyAWSRetryer with sensible defaults
+func DefaultConvoyAWSRetryer() *ConvoyAWSRetryer {
+	return &ConvoyAWSRetryer{NumMaxRetries: 10, MinDelay: 200, MinThrottleDelay: 500, MaxDelay: 30000}
+}
+
+// MaxRetries returns the number of maximum returns the service will use to make
+// an individual API request.
+func (c ConvoyAWSRetryer) MaxRetries() int {
+	return c.NumMaxRetries
+}
+
+var seededRand = rand.New(&lockedSource{src: rand.NewSource(time.Now().UnixNano())})
+
+// RetryRules returns the delay duration before retrying this request again
+func (c ConvoyAWSRetryer) RetryRules(r *request.Request) time.Duration {
+	minTime := c.MinDelay
+	throttle := c.shouldThrottle(r)
+	if throttle {
+		if delay, ok := getRetryDelay(r); ok {
+			log.Infof("Retrying with suggested delay from headers: %v", delay)
+			return delay
+		}
+
+		minTime = c.MinThrottleDelay
+	}
+
+	retryCount := r.RetryCount
+
+	delay := (1 << uint(retryCount)) * (seededRand.Intn(minTime) + minTime)
+
+	if delay > c.MaxDelay {
```

Comment:
could you document the relation between maxdelay and maxretries?

Reply:
done